src/components/swap/ConfirmSwapModal.tsx

```diff
@@ -46,86 +68,197 @@ export default function ConfirmSwapModal({
+  const [confirmModalState, setConfirmModalState] = useState<ConfirmModalState>(ConfirmModalState.REVIEWING)
+  const [pendingModalSteps, setPendingModalSteps] = useState<PendingConfirmModalState[]>([])
+
   const showAcceptChanges = useMemo(
-    () => Boolean(trade && originalTrade && tradeMeaningfullyDiffers(trade, originalTrade)),
-    [originalTrade, trade]
+    () =>
+      Boolean(
+        trade &&
+          originalTrade &&
+          tradeMeaningfullyDiffers(trade, originalTrade) &&
+          confirmModalState !== ConfirmModalState.PENDING_CONFIRMATION
+      ),
+    [confirmModalState, originalTrade, trade]
   )
 
+  const { chainId } = useWeb3React()
+  const trace = useTrace()
+  const maximumAmountIn = useMaxAmountIn(trade, allowedSlippage)
+  const [approvalError, setApprovalError] = useState<PendingModalError>()
+
+  const confirmed = useIsTransactionConfirmed(txHash)
+
+  useEffect(() => {
+    if (!showAcceptChanges && isAllowing(confirmModalState) && allowance.state === AllowanceState.ALLOWED) {
+      onConfirm()
+      setConfirmModalState(ConfirmModalState.PENDING_CONFIRMATION)
+    }
+  }, [allowance, confirmModalState, onConfirm, pendingModalSteps.length, showAcceptChanges])
+
+  const previousPermitNeeded = usePrevious(
+    allowance.state === AllowanceState.REQUIRED ? allowance.needsPermit2Approval : undefined
+  )
+  useEffect(() => {
+    async function requestSignature() {
+      // We successfully requested Permit2 approval and need to move to the signature step.
+      try {
+        if (allowance.state === AllowanceState.REQUIRED) {
+          setConfirmModalState(ConfirmModalState.PERMITTING)
+          await allowance.permit()
+        }
+      } catch (e) {
+        setConfirmModalState(ConfirmModalState.REVIEWING)
+        if (didUserReject(e)) {
+          return
+        }
+        console.error(e)
+        setApprovalError(PendingModalError.TOKEN_APPROVAL_ERROR)
+      }
+    }
+    if (
+      allowance.state === AllowanceState.REQUIRED &&
+      allowance.needsSignature &&
+      // These two lines capture the state update that should trigger the signature request:
+      !allowance.needsPermit2Approval &&
+      previousPermitNeeded
+    ) {
+      requestSignature()
+    }
+  }, [allowance, previousPermitNeeded])
+
+  const updateAllowance = useCallback(async () => {
+    invariant(allowance.state === AllowanceState.REQUIRED)
+    try {
+      if (allowance.needsPermit2Approval) {
+        setConfirmModalState(ConfirmModalState.APPROVING_TOKEN)
+        await allowance.approve()
+      } else {
+        setConfirmModalState(ConfirmModalState.PERMITTING)
+        await allowance.permit()
+      }
+      sendAnalyticsEvent(InterfaceEventName.APPROVE_TOKEN_TXN_SUBMITTED, {
+        chain_id: chainId,
+        token_symbol: maximumAmountIn?.currency.symbol,
+        token_address: maximumAmountIn?.currency.address,
+        ...trace,
+      })
```

Comment:
This will send an APPROVE_TOKEN_TXN_SUBMITTED event even when only a permit signature occurred

Reply:
updated